File: sociallang/lang/lexer.py
from __future__ import annotations

from dataclasses import dataclass

KEYWORDS = {
    "sim", "agents", "memory", "role", "fn", "phase", "win_condition", "loop",
    "let", "if", "else", "while", "for", "in", "return", "run", "break",
    "true", "false", "null", "and", "or", "not",
}

# Longest-match-first so e.g. ".." isn't tokenized as two "." tokens.
SYMBOLS = [
    "..", "==", "!=", "<=", ">=",
    "{", "}", "(", ")", "[", "]", ",", ":", ".", "+", "-", "*", "/", "%",
    "<", ">", "=",
]


@dataclass
class Token:
    kind: str  # IDENT | NUMBER | STRING | KEYWORD | SYMBOL | EOF
    value: str
    line: int


class LexError(Exception):
    pass


def tokenize(source: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    line = 1
    n = len(source)

    while i < n:
        ch = source[i]

        if ch == "\n":
            line += 1
            i += 1
            continue
        if ch in " \t\r":
            i += 1
            continue
        if ch == "/" and i + 1 < n and source[i + 1] == "/":
            while i < n and source[i] != "\n":
                i += 1
            continue

        if ch == '"':
            j = i + 1
            buf = []
            while j < n and source[j] != '"':
                if source[j] == "\\" and j + 1 < n:
                    esc = source[j + 1]
                    buf.append({"n": "\n", "t": "\t", '"': '"', "\\": "\\"}.get(esc, esc))
                    j += 2
                    continue
                buf.append(source[j])
                j += 1
            if j >= n:
                raise LexError(f"unterminated string starting at line {line}")
            tokens.append(Token("STRING", "".join(buf), line))
            i = j + 1
            continue

        if ch.isdigit():
            j = i
            while j < n and source[j].isdigit():
                j += 1
            # A single '.' followed by a digit is a decimal point; '..' (the range
            # operator, e.g. "agents: 6..10") must NOT be swallowed into the number.
            if j < n and source[j] == "." and j + 1 < n and source[j + 1].isdigit():
                j += 1
                while j < n and source[j].isdigit():
                    j += 1
            tokens.append(Token("NUMBER", source[i:j], line))
            i = j
            continue

        if ch.isalpha() or ch == "_":
            j = i
            while j < n and (source[j].isalnum() or source[j] == "_"):
                j += 1
            word = source[i:j]
            tokens.append(Token("KEYWORD" if word in KEYWORDS else "IDENT", word, line))
            i = j
            continue

        matched = next((s for s in SYMBOLS if source.startswith(s, i)), None)
        if matched:
            tokens.append(Token("SYMBOL", matched, line))
            i += len(matched)
            continue

        raise LexError(f"unexpected character {ch!r} at line {line}")

    tokens.append(Token("EOF", "", line))
    return tokens

File: cpp/src/engine/lexer.cpp
#include "engine/lexer.h"

#include <array>
#include <string_view>
#include <unordered_set>

#include "engine/ast.h"

namespace sl {

namespace {

const std::unordered_set<std::string_view> kKeywords = {
    "sim", "agents", "memory", "role", "fn", "phase", "win_condition", "loop", "world",
    "let", "if", "else", "while", "for", "in", "return", "run", "break",
    "true", "false", "null", "and", "or", "not",
};

// Two-character symbols first, so ".." wins over ".".
constexpr std::array<std::string_view, 22> kSymbols = {
    "..", "==", "!=", "<=", ">=", "{", "}", "(", ")", "[", "]",
    ",", ":", ".", "+", "-", "*", "/", "%", "<", ">", "=",
};

bool isDigit(char c) { return c >= '0' && c <= '9'; }
bool isAlpha(char c) { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'); }

}  // namespace

std::vector<Token> tokenize(const std::string& src) {
  std::vector<Token> tokens;
  tokens.reserve(src.size() / 3);
  size_t i = 0, n = src.size();
  int line = 1;

  while (i < n) {
    char ch = src[i];
    if (ch == '\n') { ++line; ++i; continue; }
    if (ch == ' ' || ch == '\t' || ch == '\r') { ++i; continue; }
    if (ch == '/' && i + 1 < n && src[i + 1] == '/') {
      while (i < n && src[i] != '\n') ++i;
      continue;
    }
    if (ch == '"') {
      size_t j = i + 1;
      std::string buf;
      while (j < n && src[j] != '"') {
        if (src[j] == '\\' && j + 1 < n) {
          char esc = src[j + 1];
          buf += esc == 'n' ? '\n' : esc == 't' ? '\t' : esc;
          j += 2;
          continue;
        }
        if (src[j] == '\n') ++line;
        buf += src[j++];
      }
      if (j >= n) throw ParseError("line " + std::to_string(line) + ": unterminated string");
      tokens.push_back({TK::String, std::move(buf), line});
      i = j + 1;
      continue;
    }
    if (isDigit(ch)) {
      size_t j = i;
      while (j < n && isDigit(src[j])) ++j;
      if (j + 1 < n && src[j] == '.' && isDigit(src[j + 1])) {
        ++j;
        while (j < n && isDigit(src[j])) ++j;
      }
      tokens.push_back({TK::Number, src.substr(i, j - i), line});
      i = j;
      continue;
    }
    if (isAlpha(ch) || ch == '_') {
      size_t j = i;
      while (j < n && (isAlpha(src[j]) || isDigit(src[j]) || src[j] == '_')) ++j;
      std::string word = src.substr(i, j - i);
      TK kind = kKeywords.count(word) ? TK::Keyword : TK::Ident;
      tokens.push_back({kind, std::move(word), line});
      i = j;
      continue;
    }
    bool matched = false;
    for (auto sym : kSymbols) {
      if (src.compare(i, sym.size(), sym) == 0) {
        tokens.push_back({TK::Symbol, std::string(sym), line});
        i += sym.size();
        matched = true;
        break;
      }
    }
    if (matched) continue;
    throw ParseError("line " + std::to_string(line) + ": unexpected character '" + std::string(1, ch) + "'");
  }
  tokens.push_back({TK::End, "", line});
  return tokens;
}

}  // namespace sl
